modules/storage-rocksdb/src/main/java/org/apache/ignite/internal/storage/rocksdb/RocksDbMvPartitionStorage.java

```diff
@@ -0,0 +1,494 @@
+                    // in this code. Now let's descrie what we'll find, assuming that iterator found something:
+                    //  - if timestamp is null:
+                    //      - this seek will find the newest version of the next row in iterator. Exactly what we need.
+                    //  - if timestamp is not null:
+                    //      - suppose that seek key buffer has the following value: "| P0 | R0 | T0 |" (partition, row id, timestamp)
+                    //        and iterator finds something, let's denote it as "| P0 | R1 | T1 |" (partition must match). Again, there are
+                    //        few possibilities here:
+                    //          - R1 == R0, this means a match. By the rules of ordering we derive that T1 >= T0. Timestamps are stored in
+                    //            descending order, this means that we found exactly what's needed.
+                    //          - R1 > R0, this means that we found next row and T1 is either missing (pending row) or represents the latest
+                    //            version of the row. It doesn't matter in this case, because this row id will be reused to find its value
+                    //            at time T0. Additional "seek" will be required to do it.
+                    it.seek(seekKeyBuf.array());
+
+                    // Finish scan if nothing was found.
+                    if (invalid(it)) {
+                        return false;
+                    }
+
+                    // Read the actual key into a direct buffer.
+                    it.key(directBuffer);
+
+                    // To understand this condition please read that huge comment above.
+                    if (timestamp == null || !directBuffer.equals(seekKeyBufSliceWithoutTimestamp)) {
+                        // Copy actual row id into a "seekKeyBuf" buffer.
+                        GridUnsafe.copyMemory(
+                                null, GridUnsafe.bufferAddress(directBuffer) + ROW_ID_OFFSET,
+                                seekKeyBuf.array(), GridUnsafe.BYTE_ARR_OFF + ROW_ID_OFFSET,
+                                igniteRowIdSize
+                        );
+
+                        // Perform additional "seek" if timestamp is not null. Motivation for it is described in comments above.
+                        if (timestamp != null) {
+                            // At this point, "seekKeyBuf" has row id that exists in partition.
+                            it.seek(seekKeyBuf.array());
+
+                            // Iterator may not be valid if that row was created after required timestamp.
+                            if (invalid(it)) {
+                                return false;
+                            }
+
+                            // Or iterator may still be valid even if there's no version for required timestamp. In this case row id
+                            // itself will be different and must check it.
```

Comment:
'and WE must check it', otherwise, it reads is if row id had to check it

Reply:
Sure